File: implement-shell-tools/wc/wc.mjs
import { program } from "commander";

import { promises as fs } from "node:fs";

program
  .name("wc")
  .description("wc implementation")
  .argument("<paths...>", "the file path to process")
  .option("-l", "count  lines")
  .option("-w", "count  words")
  .option("-c", "count characters");

program.parse();

const paths = program.args;
const options = program.opts();
const noFlag = !options.l && !options.w && !options.c;

const total = {};
let hadError = false;
for (const path of paths) {
  try {
    const content = await fs.readFile(path, "utf-8");

    const linesCounter = content.split("\n").length - 1;
    const trimmedContent = content.trim();
    const wordsCounter =
      trimmedContent === "" ? 0 : trimmedContent.split(/\s+/).length;
    const characterCounter = content.length;

    const results = [];
    if (options.l || noFlag) {
      results.push(linesCounter);
      total["lineCounter"] = (total["lineCounter"] ?? 0) + linesCounter;
    }
    if (options.w || noFlag) {
      results.push(wordsCounter);
      total["wordsCounter"] = (total["wordsCounter"] ?? 0) + wordsCounter;
    }
    if (options.c || noFlag) {
      results.push(characterCounter);
      total["characterCounter"] =
        (total["characterCounter"] ?? 0) + characterCounter;
    }

     console.log(results.map(value=> String(value).padStart(4)).join(" ") +"    "+ path)
  
  } catch (error) {
    console.error(error.message);
    hadError = true;
  }
}
if (paths.length > 1) {

  console.log(Object.values(total).map(value=> String(value).padStart(4)).join(" "),"total")
  
}
if (hadError) {
  process.exitCode = 1;
}

File: implement-shell-tools/wc/wc.py
import argparse

parser = argparse.ArgumentParser(
    prog="wc", description="Count lines, words, and characters in one or more files."
)

parser.add_argument("-c", action="store_true", help="Count the number of characters.")
parser.add_argument("-w", action="store_true", help="Count the number of words.")
parser.add_argument("-l", action="store_true", help="Count the number of lines.")
parser.add_argument("paths", nargs="+", help="Path(s) to the file(s) to process.")


total_results = {}


args = parser.parse_args()
paths = args.paths
no_flags = not args.c and not args.w and not args.l

for path in paths:
    with open(path, "r") as f:
        content = f.read()

    count_lines = len(content.splitlines())
    count_words = len(content.split())
    count_characters = len(content)

    results = {}

    if no_flags or args.l:
        results["count_lines"] = count_lines

    if no_flags or args.w:
        results["count_words"] = count_words

    if no_flags or args.c:
        results["count_characters"] = count_characters

    for key, value in results.items():
        total_results[key] = total_results.get(key, 0) + value

    print(" ".join(map(str, results.values())), path)

if len(paths) > 1:
    print(" ".join(map(str, total_results.values())), "total")
